python/ray/services.py

```diff
@@ -505,62 +442,106 @@ def start_redis(node_ip_address,
-            cleanup=cleanup)
+        if not use_credis:
+            redis_shard_port, _ = _start_redis_instance(
+                node_ip_address=node_ip_address,
+                port=redis_shard_ports[i],
+                redis_max_clients=redis_max_clients,
+                stdout_file=redis_stdout_file,
+                stderr_file=redis_stderr_file,
+                cleanup=cleanup)
+        else:
+            assert num_redis_shards == 1, \
+                "For now, RAY_USE_NEW_GCS supports 1 shard, and credis "\
+                "supports 1-node chain for that shard only."
+            redis_shard_port, _ = _start_redis_instance(
+                node_ip_address=node_ip_address,
+                port=redis_shard_ports[i],
+                redis_max_clients=redis_max_clients,
+                stdout_file=redis_stdout_file,
+                stderr_file=redis_stderr_file,
+                cleanup=cleanup,
+                executable=CREDIS_EXECUTABLE,
+                # It is important to load the credis module BEFORE the ray
+                # module, as the latter contains an extern declaration that the
+                # former supplies.
+                modules=[CREDIS_MEMBER_MODULE, REDIS_MODULE])
+
         if redis_shard_ports[i] is not None:
             assert redis_shard_port == redis_shard_ports[i]
         shard_address = address(node_ip_address, redis_shard_port)
         redis_shards.append(shard_address)
         # Store redis shard information in the primary redis shard.
-        redis_client.rpush("RedisShards", shard_address)
+        primary_redis_client.rpush("RedisShards", shard_address)
+
+    if use_credis:
+        # Configure the chain state.
+        primary_redis_client.execute_command("MASTER.ADD", node_ip_address,
+                                             redis_shard_port)
+        shard_client = redis.StrictRedis(
+            host=node_ip_address, port=redis_shard_port)
+        shard_client.execute_command("MEMBER.CONNECT_TO_MASTER",
+                                     node_ip_address, port)
 
     return redis_address, redis_shards
 
 
-def start_redis_instance(node_ip_address="127.0.0.1",
-                         port=None,
-                         redis_max_clients=None,
-                         num_retries=20,
-                         stdout_file=None,
-                         stderr_file=None,
-                         cleanup=True,
-                         executable=REDIS_EXECUTABLE,
-                         module=REDIS_MODULE):
+def _start_redis_instance(node_ip_address="127.0.0.1",
+                          port=None,
+                          redis_max_clients=None,
+                          num_retries=20,
+                          stdout_file=None,
+                          stderr_file=None,
+                          cleanup=True,
+                          executable=REDIS_EXECUTABLE,
+                          modules=[REDIS_MODULE]):
```

Comment:
let's not use mutable values for default args

Reply:
What do you mean?  Aren't all other arguments in this func mutable in some sense?

If you could give the concrete alternative you have in mind happy to look into changing

Reply:
No, e.g., the default argument for `cleanup` will always be `True`. However, if you do something like `modules.append('XXX')` inside of the function, then the default argument for `modules` will be changed to `[REDIS_MODULE, 'XXX']`. Does that make sense?

Just make it `module=None` and then inside of the function do

```python
if modules is None:
    modules = [REDIS_MODULE]
```

Reply:
The link you point to seems to be fine with `constexpr`, which allows the compiler to optimize more than `const`.

Reply:
Done